src/core/src/lib.rs

```diff
@@ -238,3 +239,107 @@ pub trait Device: Sized {
     /// Cleanup unused resources. This should be called between frames. 
     fn cleanup(&mut self);
 }
+
+
+/// An `Instance` holds per-application state for a specific backend
+pub trait Instance {
+    /// Associated `Adapter` type.
+    type Adapter: Adapter;
+    /// Associated `Surface` type.
+    type Surface: Surface;
+
+    /// Associated native `Window` type.
+    type Window;
+
+    /// Instantiate a new `Instance`, this is our entry point for applications
+    fn create() -> Self;
+
+    /// Enumerate all available adapters supporting this backend 
+    fn enumerate_adapters(&self) -> Vec<Self::Adapter>;
+
+    /// Create a new surface from a native window.
+    fn create_surface(&self, window: &Self::Window) -> Self::Surface;
+}
+
+/// Represents a physical or virtual device, which is capable of running the backend.
+pub trait Adapter {
+    /// Associated `CommandQueue` type.
+    type CommandQueue: CommandQueue;
+    /// Associated `Device` type.
+    type Device: Device;
+    /// Associated `QueueFamily` type.
+    type QueueFamily: QueueFamily;
+
+    /// Create a new device and command queues.
+    fn open<'a, I>(&self, queue_descs: I) -> (Self::Device, Vec<Self::CommandQueue>)
+        where I: Iterator<Item=(&'a Self::QueueFamily, u32)>;
+
+    /// Get the `AdapterInfo` for this adapater.
+    fn get_info(&self) -> &AdapterInfo;
+
+    /// Return the supported queue families for this adapter.
+    fn get_queue_families(&self) -> Iter<Self::QueueFamily>;
```

Comment:
I considered it easier to use to return `Iter` instead of the slice. (It's basically the iterator equivalent to &[]).
Will change it if you prefer returning a slice.

`QueueFamily` is defined as trait as it was easier to implement and wasn't sure if it can be implemented as single struct for all backends. I can try to define a struct for it. Difficulties arised because of 
checking for presentation support for a surface.